Clarify and split the "malformed" definition (#3352)
quicwg/base-drafts: draft-ietf-quic-http.md

3. optionally, trailing headers, if present (see {{!RFC7230}}, Section 4.1.2),
   sent as a single HEADERS frame.

Receipt of DATA and HEADERS frames in any other sequence MUST be treated as a
connection error of type H3_FRAME_UNEXPECTED ({{errors}}).

A server MAY send one or more PUSH_PROMISE (see {{frame-push-promise}}) or
DUPLICATE_PUSH (see {{frame-duplicate-push}}) frames before, after, or
interleaved with the frames of a response message. These PUSH_PROMISE and
DUPLICATE_PUSH frames are not part of the response; see {{server-push}} for

> more details.  These frames are not permitted in pushed responses.

Reviewer: ```suggestion
more details.  These frames MUST NOT be sent in pushed responses.
```
or

```suggestion
more details.  A pushed response is malformed if it includes PUSH_PROMISE or DUPLICATE_PUSH frames.
```

Author: Fixed, but I'm going to reverse that.  I like Kazuho's logic that misuse of a frame is a connection error, because the peer is the only one who could have generated it.  I'm going to make these H3_FRAME_UNEXPECTED instead.

> An HTTP request/response exchange fully consumes a bidirectional QUIC stream.

Reviewer 2: I guess while we are changing so much text we could make it explicity and say

"An HTTP request/response exchange fully consumes a client-initiated bidirectional QUIC stream."

Author: Good call.  Yes, this covers pushes, too -- it's a response message that follows this format.  (I don't see *why* you'd push with a non-final response in the sequence, but it's totally valid....)

Author: After reading the whole section, I think that while the section intends to talk about both push and requests, this paragraph is specific to requests.  I've added more discussion of push streams above, but I think I'm going to leave this particular paragraph alone, modulo fixing @LPardue's initial suggestion.

or response containing uppercase header field names MUST be treated as
malformed ({{malformed}}).

Like HTTP/2, HTTP/3 does not use the Connection header field to indicate

> connection-specific header fields; in this protocol, connection- specific

Reviewer 2: ```suggestion
connection-specific header fields; in this protocol, connection-specific
``` or is the space intended?

need to remove any header fields nominated by the Connection header field, along
with the Connection header field itself.  Such intermediaries SHOULD also remove
other connection-specific header fields, such as Keep-Alive, Proxy-Connection,
Transfer-Encoding, and Upgrade, even if they are not nominated by the Connection
header field.

#### Pseudo-Header Fields

As in HTTP/2, HTTP/3 uses special pseudo-header fields beginning with the ':'
character (ASCII 0x3a) to convey the target URI, the method of the request, and
the status code for the response.

Pseudo-header fields are not HTTP header fields.  Endpoints MUST NOT generate
pseudo-header fields other than those defined in this document, except as
negotiated via an extension; see {{extensions}}.

Pseudo-header fields are only valid in the context in which they are defined.
Pseudo-header fields defined for requests MUST NOT appear in responses;
pseudo-header fields defined for responses MUST NOT appear in requests.
Pseudo-header fields MUST NOT appear in trailers.  Endpoints MUST treat a
request or response that contains undefined or invalid pseudo-header fields as

> malformed ({{malformed}}).

Author: This probably needs to be forcibly rewrapped -- obviously the editor's auto-wrapping was confused by the spacing.

Author: Nope, it actually does wrap this way by sheer coincidence.  (Or @martinthomson was playing a really long prank in RFC 7540.)

    translating from an HTTP/1.1 request that has a request target in
    origin or asterisk form (see [RFC7230], Section 5.3).  Clients
    that generate HTTP/3 requests directly SHOULD use the ":authority"
    pseudo-header field instead of the Host header field.  An
    intermediary that converts an HTTP/3 request to HTTP/1.1 MUST
    create a Host header field if one is not present in a request by
    copying the value of the ":authority" pseudo-header field.

  ":path":

  : Contains the path and query parts of the target URI (the "path-absolute"
    production and optionally a '?' character followed by the "query"
    production (see Sections 3.3 and 3.4 of [RFC3986]).  A request in asterisk
    form includes the value '*' for the ":path" pseudo-header field.

  : This pseudo-header field MUST NOT be empty for "http" or "https"
    URIs; "http" or "https" URIs that do not contain a path component
    MUST include a value of '/'.  The exception to this rule is an
    OPTIONS request for an "http" or "https" URI that does not include
    a path component; these MUST include a ":path" pseudo-header field
    with a value of '*' (see [RFC7230], Section 5.3.4).

> All HTTP/3 requests MUST include exactly one valid value for the ":method",

Author: Arguably, this language from 7540 permits it to also contain any number of invalid values.  Worth rephrasing.

    accurately, this pseudo-header field MUST be omitted when
    translating from an HTTP/1.1 request that has a request target in
    origin or asterisk form (see [RFC7230], Section 5.3).  Clients
    that generate HTTP/3 requests directly SHOULD use the ":authority"
    pseudo-header field instead of the Host header field.  An
    intermediary that converts an HTTP/3 request to HTTP/1.1 MUST
    create a Host header field if one is not present in a request by
    copying the value of the ":authority" pseudo-header field.

  ":path":

  : Contains the path and query parts of the target URI (the "path-absolute"
    production and optionally a '?' character followed by the "query"
    production (see Sections 3.3 and 3.4 of [RFC3986]).  A request in asterisk
    form includes the value '*' for the ":path" pseudo-header field.

  : This pseudo-header field MUST NOT be empty for "http" or "https"
    URIs; "http" or "https" URIs that do not contain a path component
    MUST include a value of '/'.  The exception to this rule is an
    OPTIONS request for an "http" or "https" URI that does not include
    a path component; these MUST include a ":path" pseudo-header field

>     with a value of '*' (see [RFC7230], Section 5.3.4).

Reviewer: I think that we should use "Section X.X of [RFCblah]" as the one form of external section references.

Author: Fixing the ones in this imported text, but there are enough other instances of that pattern to handle separately.

All HTTP/3 requests MUST include exactly one value for the ":method", ":scheme",
and ":path" pseudo-header fields, unless it is a CONNECT request ({{connect}}).
An HTTP request that omits mandatory pseudo-header fields or contains invalid
values for those fields is malformed ({{malformed}}).

HTTP/3 does not define a way to carry the version identifier that is included in
the HTTP/1.1 request line.

For responses, a single ":status" pseudo-header field is defined that carries
the HTTP status code field (see [RFC7231], Section 6).  This pseudo-header field
MUST be included in all responses; otherwise, the response is malformed (Section

> 8.1.2.6).

Reviewer: Fix this section reference.

### Malformed Requests and Responses {#malformed}

A malformed request or response is one that is an otherwise valid sequence of
frames but is invalid due to the presence of prohibited header fields, the
absence of mandatory header fields, an invalid sequence of HTTP messages, or the

> inclusion of uppercase header field names.

Reviewer 2: You don't mention pseudo-header fields in this paragraph but I think it would help to do so (because earlier it says `Pseudo-header fields are not HTTP header fields.`).

### Malformed Requests and Responses {#malformed}

A malformed request or response is one that is an otherwise valid sequence of
frames but is invalid due to:

- the presence of prohibited header fields or pseudo-header fields
- the absence of mandatory pseudo-header fields

> - invalid values for pseudo-header fields

Reviewer: If this is intended to be exhaustive, then you need to include ordering of pseudo-header fields and header fields.

And maybe the inclusion of a non-Final status code after a final status code.

Author: > If this is intended to be exhaustive, then you need to include ordering of pseudo-header fields and header fields.

Good call.

> And maybe the inclusion of a non-Final status code after a final status code.

This would be one instance of an invalid sequence of messages.
